feat(rpc): add `getaddressdeltas` endpoint

dorianvp opened the pull request:
## Motivation

Closes #197.

## Solution

Implement `getaddressdeltas`.

### PR Checklist

<!-- Check as many boxes as possible. -->

- [x] The PR name is suitable for the release notes.
- [x] The solution is tested.
- [x] The documentation is up to date.

nachog00 committed:
Add `get_address_deltas` RPC method (barebones)
- Modified `connector.rs` to add the `get_address_deltas()` async method stub
- Updated `response.rs` to include `GetAddressDeltasResponse` struct and associated error definition
- Imported `GetAddressDeltasResponse` in `fetch.rs` and `state.rs`, adding async function for `get_address_deltas`
- Updated `indexer.rs` to define the `get_address_deltas()` method in the `ZcashIndexer` trait

Additions provide functionality for retrieving address delta changes via RPC, aligning with the `zcashd` implementation.

nachog00 committed:
Implement AddressDelta and refactor get_address_deltas method
- Modified `zaino-fetch/src/jsonrpsee/response.rs` to introduce `AddressDelta` struct and include it in `GetAddressDeltasResponse`
- Updated `zaino-state/src/backends/fetch.rs` and `indexer.rs` to include `request` parameter in `get_address_deltas` method
- Adjusted async function in `zaino-state/src/backends/state.rs` to handle building of `GetAddressDeltasResponse` with `AddressDelta` logic
- Streamlined transaction fetching and processing, ensuring mapped results are collected appropriately for `GetAddressDeltasResponse`

nachog00 committed:
Merge branch 'dev' into getaddresdeltas

nachog00 committed:
Specify transaction object type in state.rs
- Updated `txs_result` to use `Result<Vec<Box<TransactionObject>>, Self::Error>`

nachog00 committed:
Refactor transaction filtering logic in state.rs
- Changed `map` to `try_filter_map` to streamline filtering and transformation
- Updated comments for clarity on filtering logic
- Ensured only transaction objects are collected, raw transactions are filtered out

This refactoring improves readability and clarity by combining transformation and filtering in one operation.

nachog00 committed:
Add AddressDelta calculation for transactions
- Updated `zaino-fetch/src/jsonrpsee/response.rs` to include `transaction::{Input, Output}` in imports
- Implemented `AddressDelta::from_input` and `AddressDelta::from_output` to calculate deltas from transaction data
- Updated `zaino-state/src/backends/state.rs` to extract addresses from requests for delta calculation
- Aggregated input and output deltas for transactions to compute overall address deltas

Enhancement made to track address transactions for better accuracy in balance computation.

nachog00 committed:
Add GetAddressDeltasRequest and Extend GetAddressDeltasResponse

- Updated `zaino-fetch/src/jsonrpsee/response.rs`:
  - Added `BlockInfo` and `GetAddressDeltasRequest` structs to support additional request parameters
  - Updated `GetAddressDeltasResponse` to include both simple and extended format with chain info
  - Added methods for creating requests and processing transaction objects

- Modified `zaino-state/src/backends/fetch.rs`, `zaino-state/src/backends/state.rs`, `zaino-state/src/indexer.rs`:
  - Replaced `GetAddressTxIdsRequest` with `GetAddressDeltasRequest` for consistency and extended functionality
  - Updated `get_address_deltas` logic to support fetching and processing based on `chaininfo`
  - Included stubs for fetching start and end block info for chaininfo-enabled requests due to known compilation issues

nachog00 committed:
Refactor get_address_deltas for improved readability and maintainability

- Add `get_address_txs` helper method to fetch transaction objects for addresses within block range
- Add `block_info_from_height` helper method using direct state service calls for efficiency
- Simplify main `get_address_deltas` orchestrator to 7 clear lines showing business logic flow
- Remove complex stream processing and replace with straightforward sequential approach
- Follow existing codebase patterns by adding helper methods to StateServiceSubscriber impl block
- Maintain all original functionality while improving code organization and testability

nachog00 committed:
Refactor and clean up state.rs
- Renamed `get_address_txs` to `get_taddress_txs` for consistency
- Adjusted imports in `state.rs` for better readability
- Removed unnecessary blank lines to improve code clarity
- Adjusted formatting in `get_address_deltas` for better readability

nachog00 committed:
Implement parallel transaction fetching with configurable error handling

- Replace sequential transaction fetching with parallel futures::future::join_all approach
- Add fail_fast parameter to get_taddress_txs for flexible error handling modes
- Use inline self.clone() for concurrent async closures
- Implement single filter_map with conditional error propagation vs filtering
- Maintain functional programming style without manual push loops
- Expected performance improvement: 10x-1000x for multiple transaction fetches

nachog00 committed:
Refactor state.rs for clarity and consistency
- Remove trailing spaces in comments in state.rs
- Improve formatting for join_all function in state.rs for better readability
- Add comments for clarity in get_address_deltas function in state.rs

nachog00 committed:
Refactor `GetAddressDeltasResponse` construction
- Removed redundant `simple` and `with_chain_info` methods in `response.rs`
- Modified input and output processing to use more concise iterator patterns in `response.rs`
- Updated `ZcashIndexer` implementation to directly construct `WithChainInfo` and `Simple` variants in `state.rs`

These changes enhance code readability and maintainability by eliminating unnecessary methods and using concise iterator expressions.

nachog00 committed:
Remove unused functions from response.rs
- Deleted `from_transactions_simple` and `from_transactions_with_chain_info` functions
- Functions were not being used, simplifying the codebase

nachog00 committed:
Remove redundant comment in state.rs
- Deleted an outdated comment from `block_info_from_height` function
- Comment was deemed unnecessary for understanding or maintenance purposes

nachog00 committed:
Add comprehensive integration tests for get_address_deltas endpoint

- Added integration tests for both regtest and testnet environments
- Tests cover both simple response (chaininfo=false) and extended response (chaininfo=true)
- Validates response structure and ensures state service matches fetch service behavior
- Tests parallel transaction fetching and polymorphic response handling

nachog00 committed:
Refactor state_service.rs for readability and organization
- Moved import of `zaino_fetch::jsonrpsee::response` to the top of the file
- Adjusted line breaks for `assert_eq!` and `assert!` macro calls for better readability
- Reformatted `GetAddressDeltasRequest` instantiation to reduce line count

Applied cargo fmt

dorianvp committed:
Merge branch 'dev' into getaddresdeltas

dorianvp committed:
chore(`getaddressdeltas`): move things around

dorianvp committed:
chore(`getaddressdeltas`): API surface cleanup

dorianvp committed:
chore(`getaddressdeltas`): cleaup pre-fixes

dorianvp committed:
chore(`getaddressdeltas`): remove invalid tests & cleanup

dorianvp committed:
chore(`getaddressdeltas`): clippy

dorianvp committed:
chore(`getaddressdeltas`): typo

zancas committed:
Merge branch 'dev' into getaddresdeltas

nachog00 commented on integration-tests/tests/state_service.rs:
```diff
@@ -1034,7 +1035,71 @@ async fn state_service_get_address_utxos_testnet() {
+    assert_eq!(fetch_service_simple_deltas, state_service_simple_deltas);
+
+    // Test response with chain info
+    let chain_info_request =
```
Not sure what this variable name is supposed to mean...? it looks like the value is a Param type, but i might be missing something?

dorianvp replied:
Fixed

nachog00 commented on zaino-fetch/src/jsonrpsee/connector.rs:
```diff
@@ -399,6 +402,29 @@ impl JsonRpSeeConnector {
 
+    /// Returns all changes for an address.
+    ///
+    /// Returns information about all changes to the given transparent addresses within the given (inclusive)
```
given "block range"?

dorianvp committed:
chore(`getaddressdeltas`): fix typos & comments

dorianvp committed:
Merge branch 'dev' into feat/rpc-getaddressdeltas

dorianvp committed:
Merge branch 'dev' into feat/rpc-getaddressdeltas

nachog00 commented on zaino-state/src/backends/fetch.rs:
```diff
@@ -218,6 +219,28 @@ impl FetchServiceSubscriber {
+    ///
+    /// block height range, default is the full blockchain.
+    /// If start or end are not specified, they default to zero.
+    /// If start is greater than the latest block height, it's interpreted as that height.
```
this applies to start, tho it seems like it makes more sense for end... ? also, if start >= lh, and end>=start, then start-end=lh right? seems like this collapses into an edge case worth of an error?

dorianvp replied:
This comment was taken from the zcashd 6.2 docs. Both `start` and `end` are clamped down to the tip it they're above. In the case where both are `> tip`, the range ends up being `[tip, tip]`.

nachog00 commented on zaino-state/src/backends/state.rs:
```diff
@@ -875,6 +878,82 @@ impl StateServiceSubscriber {
+
+        match response {
+            ReadResponse::BlockHeader { hash, .. } => Ok(BlockInfo::new(
+                hex::encode(hash.bytes_in_display_order()),
```
This line reminds me of the problem we have with txids being reversed in some places... we probably want either our won Hash type or the BlockInfo type to implement dedicated methods to construct from "order susceptible" sources, so that we don't depend on a line on an rpc endpoint for this critical bit.

I wouldn't block on this for this PR, but maybe leave a `Todo!` mentioning the problem?

dorianvp replied:
I would instead open an issue so that it doesn't get buried with a lot of in-code nitpicks.

nachog00 commented on zaino-fetch/src/jsonrpsee/response/address_deltas.rs:
```diff
@@ -0,0 +1,523 @@
+/// Request parameters for the `getaddressdeltas` RPC method.
+#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
+#[serde(untagged)]
+pub enum GetAddressDeltasParams {
```
It seems like this type supports:
- multiple addresses + height range + chaininfo boolean flag
- single address + Implicit/no range (?) + no chaininfo 

what about: single address + range + chaininfo? i suppose that's covered by `filtered` with a single item vector?

dorianvp replied:
Yeah, that's covered by a one-item vector.

nachog00 commented on zaino-fetch/src/jsonrpsee/response/address_deltas.rs:
```diff
@@ -0,0 +1,523 @@
+/// The format depends on the `chaininfo` parameter in the request.
+#[derive(Clone, Debug, PartialEq, serde::Deserialize, serde::Serialize)]
+#[serde(untagged)]
+pub enum GetAddressDeltasResponse {
```
Love this type right here. Absolutely the way to go by typing the different answer variants to offload the logic getting the data from properly modelling this.

dorianvp committed:
Merge branch 'dev' into feat/rpc-getaddressdeltas

dorianvp committed:
chore(`getaddressdeltas`): lower block range in test

dorianvp committed:
docs(`getaddressdeltas`): more accurate `BlockInfo`

dorianvp committed:
chore(`getaddressdeltas`): expand comments & add error

nachog00 commented on zaino-state/src/backends/state.rs:
```diff
@@ -897,6 +976,75 @@ impl ZcashIndexer for StateServiceSubscriber {
+            GetAddressDeltasResponse::process_transactions_to_deltas(&transactions, &addresses);
+
+        // zcashd-like ordering
+        deltas.sort_by_key(|d| (d.height, d.block_index.unwrap_or(u32::MAX), d.index));
```
I seems like this is sorting twice, right? we probably only need one of them

nachog00 commented on zaino-fetch/src/jsonrpsee/response/address_deltas.rs:
```diff
@@ -0,0 +1,536 @@
+            })
+            .collect();
+        // zcashd-like ordering: (height ASC, blockindex ASC, index ASC)
+        deltas.sort_by_key(|d| (d.height, d.block_index.unwrap_or(u32::MAX), d.index));
```
First sort

nachog00 commented on zaino-state/src/backends/state.rs:
```diff
@@ -897,6 +976,75 @@ impl ZcashIndexer for StateServiceSubscriber {
+        // zcashd-like ordering
+        deltas.sort_by_key(|d| (d.height, d.block_index.unwrap_or(u32::MAX), d.index));
+
+        if chain_info && start > Height(0) && end > Height(0) {
```
`chain_info=true` only means that we return the context on the response, right? shouldn't the height checks be performed whatever this value is?

dorianvp replied:
According to the original implementation, whether `chainInfo` is present or not determines if the ranges also get returned. Note that it's always clamped beforehand.

nachog00 replied:
ah ok, i didn't realise that this conditional block was just to gather the start and end info 👌🏼

nachog00 commented on zaino-fetch/src/jsonrpsee/response/address_deltas.rs:
```diff
@@ -0,0 +1,536 @@
+    ) -> Vec<AddressDelta> {
+        let mut deltas: Vec<AddressDelta> = transactions
+            .iter()
+            .filter(|tx| tx.height().unwrap_or(0) > 0) // Filtering out coinbase
```
IIUC, this line filters both:
- height-less txs
- txs with height=0

Not sure which of them maps to coinbase tx? maybe it's worth it to split into 2 separate `.filter` calls, one per condition, to be clearer about what's being filtered

dorianvp replied:
Actuall, that comment is outdated, good catch

dorianvp committed:
chore(`getaddressdeltas`): remove redundant sort

dorianvp committed:
chore(`getaddressdeltas`): remove old comment

dorianvp committed:
chore(`getaddressdeltas`): remove redundant check

dorianvp committed:
chore(`getaddressdeltas`): rollback untested change

dorianvp committed:
chore(`getaddressdeltas`): add test & small fix

zancas committed:
Merge branch 'dev' into feat/rpc-getaddressdeltas

nachog00 committed:
test(`getaddressdeltas`): extract and assert on variant data

Replace simple type check with pattern matching that extracts
the variant and asserts on the actual delta values.

nachog00 committed:
test(`getaddressdeltas`): complete variant extraction for all responses

Apply pattern matching with data extraction to remaining response
assertions, adding checks for:
- Simple variant with multiple addresses
- WithChainInfo variant with start/end block validation

nachog00 committed:
test(`getaddressdeltas`): add assertions for height clamping

Extract WithChainInfo variant for start_end_clamped test and
verify that the end height is properly clamped to the actual
chain height instead of the requested value.

nachog00 committed:
test(`getaddressdeltas`): assert empty deltas for non-existent address

Extract variant for non-existent address test and verify that
the deltas array is empty while still returning valid start/end
block information.

nachog00 committed:
test(`getaddressdeltas`): refactor with constants and dividers

- Define test constants for all request parameters at the top
- Replace hardcoded values in assertions with named constants
- Add clear dividers between sub-tests for readability
- Makes test more maintainable and hackable

nachog00 committed:
test(`getaddressdeltas`): use request params in assertions

Define GetAddressDeltasParams as variables and extract their
values in assertions. This ensures assertions validate against
the actual request parameters, making the test more maintainable
and reducing duplication.

nachog00 committed:
test(`getaddressdeltas`): extract sub-tests into focused functions

Break down the monolithic test into:
- setup_address_deltas_chain(): Handles chain setup and returns context
- test_simple_address_query(): Tests basic address query
- test_filtered_start_zero(): Tests filtered query with start=0
- test_filtered_with_chaininfo(): Tests WithChainInfo variant
- test_height_clamping(): Tests end height clamping behavior
- test_non_existent_address(): Tests empty results for unknown address

Benefits:
- Clean variable names (params, response) without suffix clutter
- Each function has a clear, focused purpose
- Main test is now a readable sequence of calls
- Easy to add new test cases or modify existing ones

dorianvp committed:
Merge branch 'dev' into feat/rpc-getaddressdeltas

nachog00 committed:
test(`getaddressdeltas`): organize into dedicated module

Move all address_deltas test functions into their own module
with cleaner, contextual names:
- setup_chain (was setup_address_deltas_chain)
- test_simple_query (was test_simple_address_query)
- test_with_chaininfo (was test_filtered_with_chaininfo)
- scenario (was address_deltas_scenario)

Benefits:
- Clean namespace without verbose prefixes
- Names make sense in module context
- Easy to find and maintain all related tests
- Follows project preference against mod.rs pattern

nachog00 committed:
test(`getaddressdeltas`): move test declaration to parent module

Move #[tokio::test] to parent module with implementation in
address_deltas::main(). This gives us a cleaner test name:
  zebra::get::address_deltas
instead of:
  zebra::get::address_deltas::scenario

Benefits:
- Cleaner test names in output
- Separates test declaration from implementation
- Follows common Rust pattern of module::main() entry point

nachog00 committed:
test(`getaddressdeltas`): extract module to separate file

Move address_deltas module from inline in state_service.rs to:
  integration-tests/tests/zebra/get/address_deltas.rs

Module path resolution:
- state_service.rs declares: mod zebra { mod get { mod address_deltas; } }
- Rust looks for: tests/zebra/get/address_deltas.rs (relative to test file)

Benefits:
- Cleaner state_service.rs (removed ~250 lines)
- Dedicated file for address_deltas tests
- Easier to find and modify related tests
- Follows project's preference against mod.rs pattern

nachog00 committed:
test(`getaddressdeltas`): use module-level constants

Move constants to module level instead of passing through interfaces:
- EXPECTED_TX_HEIGHT: u32 = 102
- EXPECTED_CHAIN_TIP: u32 = 104
- HEIGHT_BEYOND_TIP: u32 = 200
- NON_EXISTENT_ADDRESS: &str = "tmVqEASZxBNKFTbmASZikGa5fPLkd68iJyx"

Benefits:
- Cleaner function signatures (no expected_tx_height parameter)
- Single source of truth for all test values
- Easier to modify test constants
- No more hardcoded magic numbers throughout the code

nachog00 committed:
test(`getaddressdeltas`): use local variables for param values

Store start/end heights in local variables before constructing
params, then use those variables directly in assertions. This
eliminates the awkward pattern matching on params.

Before:
  let params = GetAddressDeltasParams::Filtered { start: 1, end: 104, ... };
  if let GetAddressDeltasParams::Filtered { start: req_start, end: req_end, .. } = &params {
    assert_eq!(response.start.height, *req_start);
  }

After:
  let start_height = 1;
  let end_height = EXPECTED_CHAIN_TIP;
  let params = GetAddressDeltasParams::Filtered { start: start_height, end: end_height, ... };
  assert_eq!(response.start.height, start_height);

Benefits:
- More ergonomic - no nested if let patterns
- Clearer - variables have descriptive names
- Simpler - direct comparison without destructuring

nachog00 committed:
test(`getaddressdeltas`): add comprehensive module documentation

Add detailed documentation explaining:

Chain Scenario:
- Block 1-100: Initial generation
- Block 101: Faucet shields funds
- Block 102: 250k zatoshis sent to recipient
- Final height: ~102

Constants & Invariants:
- EXPECTED_TX_HEIGHT (102): Transaction block height
  Invariant: Must be >= 102 based on setup

- EXPECTED_CHAIN_TIP (104): Final chain height
  Invariants: Must be >= EXPECTED_TX_HEIGHT, includes buffer

- HEIGHT_BEYOND_TIP (200): Tests height clamping
  Invariant: Must be > EXPECTED_CHAIN_TIP

- NON_EXISTENT_ADDRESS: Tests empty result handling

Modifying Parameters:
Documents implications of changing setup parameters and
how constants relate to each other.

This makes the test module self-documenting and easier to
maintain when chain setup needs to change.

nachog00 committed:
Merge branch 'feat/rpc-getaddressdeltas' of github.com:zingolabs/zaino into feat/rpc-getaddressdeltas

dorianvp commented:
I see deprecated usages... let me fix that!

dorianvp committed:
fix(`getaddressdeltas`): add deprecated markers

dorianvp commented:
> Is this mergeable?

Yeah, needs an approval.
